src/lib.rs

```diff
@@ -41,7 +50,8 @@ impl Connection {
 
     pub fn initialize(
         &self,
-        server_capabilities: serde_json::Value,
+        server_info: Option<ServerInfo>,
+        server_caps: impl Fn(&serde_json::Value) -> serde_json::Value,
```

Comment:
Hm, no, I think this is a framework in the making, because acceping a lambda is a tiny form of inversion of control. 

I think we should do this:

```rust
pub fn initialize_start() -> Value;
pub fn initialize_finish(caps: Value);
pub fn initialize(caps: Value) {
  let res = initialize_start();
  initialize_finish(caps);
}
```

Reply:
It would be more annoying to the caller to call start/finish in a sequence, but also more flexible and easier to understand.